### ISSUE TITLE
Xarray-to-vtk Prototypes 

### ISSUE THREAD

**Issue by johnkit** [Kitware/pan3d#106]
Using vtkNetCDFCFReader (sp?)

- [ ] to netcdf file to paraview
- [ ] to netcdf file to Python (PyVista Plotter for display)
- [ ] Xarray direct (no netcdf file)
- [ ] Xarray slice feature

**Comment by danlipsa** [comment]
to netcdf file to paraview
--------------------------------
File > Open and choose NetCDF Reader
Attached is a paraview state file for opening air_temperature.nc also attached.
[air_temperature.pvsm.zip](https://github.com/user-attachments/files/16742401/air_temperature.pvsm.zip)
[air_temperature.nc.zip](https://github.com/user-attachments/files/16742402/air_temperature.nc.zip)

**Comment by danlipsa** [comment]
to netcdf file to Python (PyVista Plotter for display)
-------------------------------------------------------------------
Build VTK from:
https://gitlab.kitware.com/danlipsa/vtk/-/commits/xarray_quick
then execute the attached script like this:
```PYTHONPATH=~/projects/pyvista:~/projects/vtk/xarray/build/lib/python3.10/site-packages/ python test.py```
[test.py.zip](https://github.com/user-attachments/files/16742430/test.py.zip)

**Comment by johnkit** [comment]
Interim example:

![image](https://github.com/user-attachments/assets/68b90662-d67a-43d8-ab22-584ebe615444)

**Comment by johnkit** [comment]
(also in #99)
I put a small zarr example (example.zarr.tgz) at https://data.kitware.com/#item/66d225af9eee4150438e24ae.  (Github wouldn't let me upload it here for some reason.) Note that you have to untar before trying to load :)

![image](https://github.com/user-attachments/assets/1eb1fe67-b0f0-452e-8566-cfeb4fbec67e)

**Comment by danlipsa** [comment]
Here is the WIP MR for the full xarray to vtk: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/11513

**Comment by danlipsa** [comment]
The attached script works with the xarray MR: https://gitlab.kitware.com/vtk/vtk/-/merge_requests/11513
[xarray_vtk.zip](https://github.com/user-attachments/files/17369284/xarray_vtk.zip)

**Comment by danlipsa** [comment]
![image](https://github.com/user-attachments/assets/b0e755aa-8efb-4e47-835c-8ee84f920a06)
![image](https://github.com/user-attachments/assets/2a020fbf-5a64-4844-9e93-31e7da4b0087)
![image](https://github.com/user-attachments/assets/48cd8441-8083-410b-bf20-6d38d63bfbd4)
![image](https://github.com/user-attachments/assets/26e0a501-d4e1-4bee-9a86-23972dde94c8)
![image](https://github.com/user-attachments/assets/d0a58bc9-45dc-47f5-b4c6-1f86ccb80e14)
![image](https://github.com/user-attachments/assets/a89f393c-b2dd-45ee-8b87-9ebde44d6398)
![image](https://github.com/user-attachments/assets/0922b16a-1305-4d90-87ac-f7ee83dbf4a3)
![image](https://github.com/user-attachments/assets/8f694481-40d8-432e-aa91-201af23bd77b)
![image](https://github.com/user-attachments/assets/17a71660-2ba4-445e-8b06-6f68bdec5689)

**Comment by danlipsa** [comment]
[xarray_vtk.zip](https://github.com/user-attachments/files/18156778/xarray_vtk.zip)

**Comment by jourdain** [comment]
@danlipsa are everything available in the nightly VTK wheel?

**Comment by danlipsa** [comment]
Yes, everything is there. I used: https://gitlab.kitware.com/vtk/vtk/-/jobs/10654362
for linux.

I use
https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/NetCDF/Testing/Python/GetReader.py
to either get the reader from the file directly or zero-copy from the xarray.

In few cases you'll need to:
```
            ds_tree = xr.open_datatree(args.input)
            ds_xr = ds_tree[args.node].to_dataset()
```
and then get the dataset from the xarray as before.

**Comment by jourdain** [comment]
Where did you get the __salt__ dataset?

**Comment by danlipsa** [comment]
All datasets are from: https://github.com/pydata/xarray-data/
The salt one is ROMS_example.nc

**Comment by jourdain** [comment]
Here is the API we have in our dataset builder that we would need to map to the VTK API. Some are rectilinear specific but for now I'm just listing everything so we can create an adapter that is more python friendly:

- __Rectilinear specific__
  - x, x_size, y, y_size, z, z_size (coord name + size) (? GetAllDimensions())
  - slice_extends
  - slices (set/get sub domain selection) 
- __Generic__
  - t (set/get coord name), t_index, t_size, t_labels (list time in a human fashion)
  - available_coords (list 1d xarray coords)
  - arrays (set/get the fields to load/loaded)
  - available_arrays (list available arrays => `reader.GetAllVariableArrayNames()`)

__Questions__: For the VTK reader how do I ...

1. Handle time? Use pipeline information pass?
2. Select fields to load? I can figure out, what is available, but how can I activate/deactivate specific field?
3. For compatible format, how can I query/set a subset selection (slice, range)

__Next step__:

Once we understand how do some of those operations, we should create a Python adapter or improve the C++ code to help that integration process.

__Comments__:

1. The xarray accessor should just do `ds.vtk.reader` [no ()] by defining the method as a property. [Will need to fix](https://gitlab.kitware.com/vtk/vtk/-/blob/master/IO/NetCDF/Testing/Python/GetReader.py#L21)
2. The helper `get_accessor` should either be internal or a `@property`
3. We may want another Python class that decorate an xarray into a vtk mesh algo/reader.

**Comment by danlipsa** [comment]
> **Questions**: For the VTK reader how do I ...
> 
> 1. Handle time? Use pipeline information pass?
> 2. Select fields to load? I can figure out, what is available, but how can I activate/deactivate specific field?
> 3. For compatible format, how can I query/set a subset selection (slice, range)

The reader returned is a vtkNetCDFCFReader so:
1. Indeed time is handled using the pipeline.
2. Many readers have this: https://vtk.org/doc/nightly/html/classvtkNetCDFReader.html#ab457b03b7de7908dd6437b14fcfc7538
4. vtkExtractGrid

**Comment by jourdain** [comment]
Some weird behavior that I'm noticing... When loading an XArray without enabling any field, the output is a `vtkImageData`. Once I activate a field, it became a `vtkStructuredGrid`.

I think the switch of dataset type is messing up the VTK pipeline. I'm going to talk with Berk about that and may implement a work around into our viewers/explorers.

**Comment by danlipsa** [comment]
Try to load ROMS_example.nc in paraview. You'll see that you can select Dimensions. (you should choose (s_rho, eta_rho, xi_rho)). The list you see is AllDimensions you asked about earlier: the list of all possible set of dimensions for all variables.

Once you select a dimension, all variables that have that dimension are loaded. Probably you'll need to do something similar in trame.

**Comment by danlipsa** [comment]
For different variables, you can have different types of output.
